### ISSUE TITLE
Torchvision Resnet 50 accuracy

### ISSUE THREAD

**Issue by Esaada** [pytorch/vision#1757]
Hey, Pytorch’s (torchvision) Resnet 50 accuracy is declared to be 76.15.
But when I’m using the training script from PyTorch’s repo, which is mentioned in the official torchvision website(https://pytorch.org/docs/stable/torchvision/models.html#classification):
[https://github.com/pytorch/examples/blob/master/imagenet/main.py]
and the Resnet50 from torchvision:
[https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py]
When training it, after one epoch I’m getting an accuracy of 76.6, how can it be? isn’t the models fully trained?

Thanks!

**Comment by fmassa** [comment]
Hi,

Can you share the code you are using to train the model? I suspect that you are passing the `--pretrained` flag to the training code, which starts the model from pre-trained weights

**Comment by BattashB** [comment]
I tried multiple ways, but the straight way is exactly the same as the script I mentioned:
https://github.com/pytorch/examples/blob/master/imagenet/main.py

You asled about the pretrained part:
`   if args.pretrained:
        print("=> using pre-trained model '{}'".format(args.arch))
        model = models.__dict__[args.arch](pretrained=True)
    else:
        print("=> creating model '{}'".format(args.arch))
        model = models.__dict__[args.arch]()
`
All I'm doing is to continue training using --lr=0.0001.

**Comment by Esaada** [comment]
I tried multiple ways, but the straight way is exactly the same as the script I mentioned:
https://github.com/pytorch/examples/blob/master/imagenet/main.py

You asled about the pretrained part:
`   if args.pretrained:
        print("=> using pre-trained model '{}'".format(args.arch))
        model = models.__dict__[args.arch](pretrained=True)
    else:
        print("=> creating model '{}'".format(args.arch))
        model = models.__dict__[args.arch]()
`
All I'm doing is to continue training using --lr=0.0001.

**Comment by fmassa** [comment]
There can be slight variations of the model accuracy during training, and this could be due to that.
Plus, if I remember correctly, when the resnet models were trained (not by me), there was a post-process step that was done to recompute the batch norm statistics. This means that doing some extra fine-tuning steps could potentially change the results a bit.

Given the above, I'm closing this issue, but let me know if you have further questions

**Comment by KaiHoo** [comment]
@fmassa Hi, I also found the problem that the reported error rates for ResNet50 and ResNet101 in the official torch vision website(https://pytorch.org/docs/stable/torchvision/models.html#classification) is higher than our re-trained results.  For example, group normalization reported a 22% top 1 error rate of  ResNet101 on ImageNet, which is 0.63% lower (I did the experiments and 22% can be achieved ). I think the results of group normalization are also trained from PyTorch?  So would you re-train the network to change the results?

Maybe the reported results from torch vision website did not use zero-gamma trick (use zero to initialize the last BN weights in each residual blocks). But the difference is not so big.

**Comment by fmassa** [comment]
@KaiHoo the resnet-pretrained models didn't use the zero-initialization trick, and we don't provide pretrained models with group-normalization in torchvision.

**Comment by KaiHoo** [comment]
@fmassa Sorry I did not make myself clear. The reported ResNet101 error rate **with Batch Normalization** in the GN paper is 22.0% (https://arxiv.org/pdf/1803.08494.pdf page 6 _Deeper models_. They use a batch size of 32 per GPU and 8 GPUs. All other training settings are the same as shown in the torchvision training scripts.)

As a reference,  _Mixup: Beyond empirical risk minimization_ also shows that ResNet101 **with Batch Normalization and without extra regularization** can achieve a 22.0% error rate. 

I understand the resnet-pretrained models didn't use the zero-initialization trick. But the difference should not be so big (0.64%). I believe it is beyond statistical significance.

**Comment by fmassa** [comment]
Performance of resnet has gradually improved over time. The first officially-reported pre-trained weights for ImageNet for ResNet101 had a top1 error of 23.6, see https://github.com/KaimingHe/deep-residual-networks

So the zero-initialization trick (which was used in the GN paper) can give (on ResNet50) 0.5 top1 accuracy, see https://github.com/pytorch/vision/pull/498#issuecomment-398436396

**Comment by KaiyuYue** [comment]
Hi @fmassa,

I want to ask that which training type did you use for training ResNet on ImageNet - DP or DDP in [examples/imagenet/main.py](https://github.com/pytorch/examples/tree/master/imagenet) ?

Thanks.

**Comment by YYangZiXin** [comment]
I meet same problem as @KaiyuYue ,  for ResNet-50 I only got 75.806% with default settings , how to reproduce the torchvision.models performances ? @fmassa

**Comment by fmassa** [comment]
@KaiyuYue @YYangZiXin IIRC the pre-trained models available in torchvision have the batch-norm statics recomputed after training is over, which gives slightly better accuracies. Those models have been trained a long time ago, and I'm afraid the code for doing the batch norm stats recomputation is lost, but it shouldn't be too hard to implement it.

**Comment by hukkai** [comment]
@fmassa Could you give some details about the the re-computation? Is it computed over the whole dataset? Or it is achieved by forwarding with very large batch size? Thanks!

**Comment by fmassa** [comment]
@hukkai I haven't done it myself, but I believe it was done over the whole train dataset.
